mqtt/mqtt3/src/proto/deserialize.rs

```diff
@@ -0,0 +1,140 @@
+#![allow(dead_code)]
+#![allow(unused_imports)]
+
+use crate::proto::packet::Publication;
+use bytes::Bytes;
+use serde::de::{self, Deserialize, DeserializeSeed, Deserializer, SeqAccess, Visitor};
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::sync::Mutex;
+
+pub struct DeserializeState {
+    payload_de_duper: PayloadDeDuper,
+}
+
+impl DeserializeState {
+    pub fn new() -> Self {
+        Self {
+            payload_de_duper: PayloadDeDuper::new(),
+        }
+    }
+}
+
+pub struct SeededDeserializer<'a, T>(&'a mut DeserializeState, PhantomData<T>);
+pub struct SeededVisitor<'a, T>(&'a mut DeserializeState, PhantomData<T>);
+
+impl<'de, 'a> DeserializeSeed<'de> for SeededDeserializer<'a, Publication> {
+    type Value = Publication;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        impl<'de, 'a> Visitor<'de> for SeededVisitor<'a, Publication> {
+            type Value = Publication;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(formatter, "an array of integers")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let topic_name = seq
+                    .next_element()?
```

Comment:
does serde provide any guarantees on the order of elements? I mean, can they have different layout depending on machine, OS, or some other random factor.

Reply:
It's based on the serialize function. If we do this, we should probably also implement the mirroring serialize functions

Reply:
What the doc says about ordering?

Reply:
https://serde.rs/deserialize-struct.html this example doesn't make it clear why it implements both `visit_seq` and `visit_map` for the same struct.